compiler/rustc_attr/src/builtin.rs

```diff
@@ -154,14 +155,27 @@ pub struct ConstStability {
 }
 
 /// The available stability levels.
-#[derive(Encodable, Decodable, PartialEq, PartialOrd, Copy, Clone, Debug, Eq, Hash)]
+#[derive(Encodable, Decodable, PartialEq, Copy, Clone, Debug, Eq, Hash)]
 #[derive(HashStable_Generic)]
 pub enum StabilityLevel {
     // Reason for the current stability level and the relevant rust-lang issue
     Unstable { reason: Option<Symbol>, issue: Option<NonZeroU32>, is_soft: bool },
     Stable { since: Symbol },
 }
 
+impl cmp::PartialOrd for StabilityLevel {
+    // This only take into account stability, not any fields.
+    // Therefore it is only `PartialOrd` and not `Ord`.
+    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+        match (self, other) {
+            (Self::Unstable { .. }, Self::Unstable { .. }) => Some(cmp::Ordering::Equal),
+            (Self::Stable { .. }, Self::Stable { .. }) => Some(cmp::Ordering::Equal),
+            (Self::Unstable { .. }, Self::Stable { .. }) => Some(cmp::Ordering::Less),
+            (Self::Stable { .. }, Self::Unstable { .. }) => Some(cmp::Ordering::Greater),
```

Comment:
I'd keep this bit in rustdoc (and remove the `PartialOrd` impl from `StabilityLevel`?).
I don't know what rustdoc rendering wants when sorting by stability, but it probably isn't something generally applicable.

Reply:
Also this can be simplified to `mem::discriminant(self).cmp(mem::discriminant(other))`.

Reply:
> Also this can be simplified to mem::discriminant(self).cmp(mem::discriminant(other)).

Does rustc guarantee that the order of the enums matches the order of the discriminants? I thought layout was unspecified.

Reply:
Yes, discriminant values are specified, see https://doc.rust-lang.org/reference/items/enumerations.html?highlight=discriminant#custom-discriminant-values-for-fieldless-enumerations for enums without data, and https://github.com/rust-lang/rfcs/blob/master/text/2363-arbitrary-enum-discriminant.md#semantics for enums with data.

Reply:
TIL, thanks!